[PyROOT] Better way to make C++ namespaces appear as Python modules

guitargeek opened the pull request:
This reimplements the mechanism to provide C++ namespace as Python
modules without replacing the built-in `__import__` function.

Credits to @chrisburr!

guitargeek commented:
The one failure on Fedora 39 is unrelated:
```
Failed
test_import (import_load_libs.ImportLoadLibs.test_import)
Test libraries loaded after importing ROOT ... ERROR

======================================================================
ERROR: test_import (import_load_libs.ImportLoadLibs.test_import)
Test libraries loaded after importing ROOT
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/github/home/ROOT-CI/src/bindings/pyroot/pythonizations/test/import_load_libs.py", line 120, in test_import
    raise Exception('Found not whitelisted libraries after importing ROOT:' \
Exception: Found not whitelisted libraries after importing ROOT:
 - libb2
 - libgomp
If the test fails with a library that is loaded on purpose, please add it to the whitelist.

----------------------------------------------------------------------
Ran 1 test in 1.405s

FAILED (errors=1)
CMake Error at /github/home/ROOT-CI/src/cmake/modules/RootTestDriver.cmake:232 (message):
  error code: 1
```

Axel-Naumann commented:
@guitargeek sure? Because we don't see this in [the nightly](https://github.com/root-project/root/actions/runs/7054741262/job/19204094244)

guitargeek commented:
I just saw Python-related failures on Fedora 39 even on a PR for 6.30 backports:
https://github.com/root-project/root/pull/14161
```
Failed
Info in <TUnixSystem::ACLiC>: creating shared library /github/home/ROOT-CI/build/roottest/python/basic/./ArgumentPassingCompiled_C.so

-- TEST COMMAND -- 
cd /github/home/ROOT-CI/build/roottest/python/basic
/usr/bin/timeout -s USR2 270s /py-venv/ROOT-CI/bin/python3.12 /github/home/ROOT-CI/src/roottest/python/basic/PyROOT_basictests.py --fixcling
-- BEGIN TEST OUTPUT --
============================= test session starts ==============================
platform linux -- Python 3.12.0, pytest-7.4.3, pluggy-1.3.0
rootdir: /github/home/ROOT-CI
plugins: anyio-4.1.0
collected 25 items

../../../../src/roottest/python/basic/PyROOT_basictests.py .........Fatal Python error: Segmentation fault

Current thread 0x00007fca5edb6740 (most recent call first):
  File "/usr/lib64/python3.12/traceback.py", line 1043 in _compute_suggestion_error
  File "/usr/lib64/python3.12/traceback.py", line 737 in __init__
  File "/usr/lib64/python3.12/traceback.py", line 159 in format_exception_only
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/_code/code.py", line 605 in exconly
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/_code/code.py", line 806 in get_exconly
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/_code/code.py", line 794 in get_source
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/_code/code.py", line 860 in repr_traceback_entry
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/_code/code.py", line 914 in repr_traceback
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/_code/code.py", line 989 in repr_excinfo
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/_code/code.py", line 701 in getrepr
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/nodes.py", line 486 in _repr_failure_py
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/python.py", line 1833 in repr_failure
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/reports.py", line 362 in from_item_and_call
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/runner.py", line 368 in pytest_runtest_makereport
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/runner.py", line 224 in call_and_report
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/runner.py", line 133 in runtestprotocol
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/runner.py", line 114 in pytest_runtest_protocol
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/main.py", line 350 in pytest_runtestloop
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/main.py", line 325 in _main
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/main.py", line 271 in wrap_session
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/main.py", line 318 in pytest_cmdline_main
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/pluggy/_callers.py", line 77 in _multicall
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/pluggy/_manager.py", line 115 in _hookexec
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/pluggy/_hooks.py", line 493 in __call__
  File "/py-venv/ROOT-CI/lib64/python3.12/site-packages/_pytest/config/__init__.py", line 169 in main
  File "/github/home/ROOT-CI/src/roottest/python/common.py", line 46 in run_pytest
  File "/github/home/ROOT-CI/src/roottest/python/basic/PyROOT_basictests.py", line 446 in <module>

Extension modules: libcppyy3_12, libcppyy, libROOTPythonizations3_12, libROOTPythonizations (total: 4)
 *** Break *** segmentation violation
/usr/bin/timeout: the monitored command dumped core

-- END TEST OUTPUT --
CMake Error at /github/home/ROOT-CI/build/RootTestDriver.cmake:186 (message):
  got exit code Segmentation fault but expected 0
```

It's not exactly the same test that fails, but it's interesting to know anyway.

Axel-Naumann commented:
That indeed different, I need to backport my "no trailing `:` in `PYTHON_PATH`" fix.

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/__init__.py:
```diff
@@ -60,7 +60,63 @@
+    def __init__(self):
+        self._module_cache = {}
+
+    def is_package(self, fullname):
```
This function sounds like a simple check, but it is actually the function that creates the ROOT module(s) and writes them to the cache. I think the two parts of the logic should be separate.

guitargeek replied:
I have simplified this by not using the caching. Implementing a cache for these recursive lookups (that should be pretty fast) actually looks like a premature optimization to me.

vepadulano reviewed (changes requested):
I believe this is really the right way forward! Although this changes just a few lines of code, it is actually a much bigger change than it might look. This involves the loading of every ROOT submodule (now we can really call them like that :) ). Some missing things that come to mind:

* The commit message of the 2nd commit should mention that we refer to the implementation of `Loader` and `MetaPathFinder` as described in https://docs.python.org/3/library/importlib.html#module-importlib.abc
* The first commit message with the cosmetic changes is probably too personal, I think leaving just the commit title suffices
* We need a test that checks attributes of the imported submodules. For example `from ROOT import RDF`, we should check that 1. the module has all the atttributes as defined at https://docs.python.org/3/library/importlib.html#importlib.abc.Loader.load_module and that it also has the correct pythonizations and they work
* We need a test that tries to import deeply nested ROOT submodules, e.g. `from ROOT import RDF.Experimental.Distributed` (which is a true Python module), `from ROOT import Experimental.RNTuple`, from `ROOT.RooFit import SomeOtherModule` etc.
* It would be nice to see better documentation in the classes that implement `Loader` and `MetaPathFinder` with references to the official Python docs and type hints.

guitargeek commented on bindings/pyroot/pythonizations/python/ROOT/__init__.py:
```diff
@@ -60,7 +60,63 @@
+from importlib.util import spec_from_loader
+
+
+class _RootNamespaceLoader(importlib.abc.Loader):
```
Thanks a lot, that's a great suggestion! I'll try it out, leaving `exec_module` a no-op.

guitargeek commented:
Thanks @vepadulano for you review! I'm in the process of addressing your comments, I only have one comment that I can't act on:

>     * We need a test that checks attributes of the imported submodules. For example `from ROOT import RDF`, we should check that 1. the module has all the atttributes as defined at https://docs.python.org/3/library/importlib.html#importlib.abc.Loader.load_module and that it also has the correct pythonizations and they work

Why is it exactly important to have all these attributes set? For the ROOT facade itself, it's easy to do (my PR does it in fact), but for arbitrary submodules I wouldn't know what to do. Often, the submodules are represented by C++ namespaces, so all these Python-module related dunder attributes are pretty meaningless. What exactly is your suggestion here?

Thanks again!
Jonas

guitargeek commented:
The test timed out, so the failure is unrelated to this PR.

vepadulano commented:
> Often, the submodules are represented by C++ namespaces, so all these Python-module related dunder attributes are pretty meaningless. What exactly is your suggestion here?

The suggestion is to make sure that we are faking well enough that we won't incur in problems down the line with the Python interpreter machinery. I admit this might be a bit pedantic from me at this point in time

vepadulano commented:
@phsft-bot build just on mac12arm/cxx20

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/__init__.py:
```diff
@@ -60,7 +60,70 @@
+from importlib.util import spec_from_loader
+
+
+def _lookup_root_module(fullname: str):
```
```suggestion
def _lookup_root_module(fullname: str) -> Optional[types.ModuleType]:
```

vepadulano replied:
Or in general I would try to find a way to represent the type returned by this function

guitargeek replied:
I found a way to have better defined types by only returning `non-None` if the attribute is actually a Python module or a C++ namespace.

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/__init__.py:
```diff
@@ -60,7 +60,70 @@
+    """
+
+    def is_package(self, fullname: str) -> bool:
+        """
+        Indicates whether the given module is considered a package.
+        We pretend to Python that everything in ROOT can be used like a package.
+        """
```
Why? So that we can chain imports like in `from ROOT.RDF.Experimental....` ?

vepadulano replied:
Maybe give a bit more explanation in the docstring

guitargeek replied:
I now explicitly check if the attribute can actually be considered a submodule, so I don't need to explain anything anymore :) (I hope :D)

vepadulano reviewed (changes requested):
Thanks a lot for the improvements! I believe we are very close now

guitargeek commented:
Hi @vepadulano, thanks for your comments! I hope my recent commit, that will be squashed before merging, is addressing your concerns.

Unfortunately, the new CI is red because I merged a PR that broke incremental builds :( But we should get feedback from Jenkins if this PR still works with my newest change.

vepadulano commented on bindings/pyroot/pythonizations/test/root_module.py:
```diff
@@ -11,48 +38,71 @@ def test_import(self):
+        if root_module_has("RooFit.Evaluator"):
+            from ROOT.RooFit import Evaluator
+
+
```
Another interesting test could be import of classes defined inside of ROOT classes, e.g.

```python
from ROOT.TDirectory import TContext
```

Does `TDirectory` qualify as a Python module in this case? As in, do we want it to qualify as one or not?

guitargeek replied:
Excellent question! With my recent update, this would not work anymore because only namespaces are considered as candidates for modules, while the previous solution that replaced `__import__` didn't care.

I think we want to make PyROOT as Pythonic as possible, and in Python, one can also not use classes as modules. For example, if you have this module:
```python
# file mymodule.py

class Foo(object):

    class Bar(object):

        pass
```
You can't do `from mymodule.Foo import Bar`:
```python
import mymodule

from mymodule import Foo
from mymodule.Foo import FooBar # This doesn't work

mymodule.Foo.Bar
```
So I'd say that no Python user would expect things like that to work to begin with.

With C++ namespaces that's a different story: there is no direct equivalent in Python, and the policy that namespaces can be used as modules is intuitive to the user.

guitargeek replied:
What is your opinion on this?

vepadulano replied:
Perfect, I agree 100% with your reply!

vepadulano commented on bindings/pyroot/pythonizations/python/ROOT/__init__.py:
```diff
@@ -30,65 +30,161 @@
+    ret = _root_facade
+    for part in keys:
+        ret = getattr(ret, part, None)
+        if ret is None:
+            return None
+        if not _can_be_module(ret):
+            return None
```
Nitpick
```suggestion
        if ret is None or not _can_be_module(ret):
            return None
```

guitargeek replied:
I'll update that

vepadulano reviewed (commented):
One nitpick and one question for discussion

guitargeek commented:
@phsft-bot build

guitargeek commented:
@phsft-bot build

dpiparo commented on bindings/pyroot/pythonizations/python/ROOT/__init__.py:
```diff
@@ -11,13 +11,13 @@
 
 # Prevent cppyy's check for the PCH
-environ['CLING_STANDARD_PCH'] = 'none'
+environ["CLING_STANDARD_PCH"] = "none"
```
Sorry to be picky, but could we maybe have a separate commit for non-functional changes such as these ones?

guitargeek replied:
Yes there is, the first commit is only style changes and the second one the functional changes

dpiparo commented on bindings/pyroot/pythonizations/python/ROOT/__init__.py:
```diff
@@ -29,58 +29,155 @@
+    # Check if the object represents a C++ namespace. Since cppyy has no
+    # dedicated Python type for C++ namespaces, we check for this using the
+    # representation of the object.
+    if repr(obj).startswith("<namespace "):
```
Does this pass through the pretty printer of the interpreter? (I think no, but triple checking b/c that would be a perf hit)

guitargeek replied:
Good question, I will investigate

guitargeek replied:
I found where it's implemented. `__repr__` doesn't call the interpreter, but CPyCppyy builds the representation from some internal flags:
https://github.com/root-project/root/blob/master/bindings/pyroot/cppyy/CPyCppyy/src/CPPScope.cxx#L175

wlav replied:
Both classes and namespaces are instances of `cppyy.types.Scope` as there's no Python reason to distinguish them. For C++ reasons to differentiate, `cppyy.reflex` exists. E.g. `obj.__cpp_reflex__(cppyy.reflex.IS_NAMESPACE)` should give the right answer. (This all in master anyway.)

guitargeek replied:
Thanks a lot! I'll keep this in mind for the next changes to the ROOT module.

dpiparo commented on bindings/pyroot/pythonizations/python/ROOT/__init__.py:
```diff
@@ -29,58 +29,155 @@
+    keys = fullname.split(".")[1:]
+    ret = _root_facade
+    for part in keys:
+        ret = getattr(ret, part, None)
```
Does this trigger in any way the Cling interpreter?

guitargeek replied:
It does, because it needs to look up things in the C++ word, like `RooFit` as an attribute of the global C++ namespaces exposed by the `ROOT` module. But there is no way around this given that indeed we want to lookup things in C++.

The same logic was there before in the old hardcoded replacement of `__import__`:
https://github.com/root-project/root/blob/master/bindings/pyroot/pythonizations/python/ROOT/_facade.py#L148

dpiparo commented on bindings/pyroot/pythonizations/python/ROOT/__init__.py:
```diff
@@ -29,58 +29,155 @@
 # Configuration for usage from Jupyter notebooks
 if _is_ipython:
     from IPython import get_ipython
+
```
Sorry again to be picky, but this is a very good candidate for a NFC commit.

guitargeek replied:
Yes, indeed it is part of the first commit that is only style changes

dpiparo commented on bindings/pyroot/pythonizations/python/ROOT/_facade.py:
```diff
@@ -28,18 +28,18 @@ def __init__(self):
 
 class _gROOTWrapper(object):
     """Internal class to manage lookups of gROOT in the facade.
```
NFC again.

guitargeek replied:
It's part of the first formatting-only commit

dpiparo commented on bindings/pyroot/pythonizations/python/ROOT/__init__.py:
```diff
@@ -29,58 +29,155 @@
-    if 'app' in facade.__dict__ and hasattr(facade.__dict__['app'], 'process_root_events'):
-        facade.__dict__['app'].keep_polling = False
-        facade.__dict__['app'].process_root_events.join()
+    if "app" in facade.__dict__ and hasattr(facade.__dict__["app"], "process_root_events"):
```
NFC

guitargeek replied:
Right, it's part of the first commit

dpiparo reviewed (changes requested):
Thanks for these changes! I have a request before approving. It seems that many, very welcome, non-functional changes were added. Could these be factored out in a specific NFC commit? This will help future readability of the history.

guitargeek commented:
Hi @dpiparo, thanks for the comments! I have addressed them by replying in the individual comments. Let me know if it's okay now

vepadulano reviewed (approved):
Thank you for the great effort and all the back-and-forth in the reviews, I believe this is good to go now, and what a nice change :D!
